### PR TITLE
Migrate from Google Analytics to Mixpanel (SCP-2326, SCP-2382)

### PR DESCRIPTION
**Pull request by eweitz**
This expands the scope of Mixpanel logging in SCP, and adds Mixpanel ports of custom-event logging previously only done in Google Analytics (GA).

Previously, Mixpanel logging was confined to advanced search.  Now, Mixpanel logging is done throughout SCP.  This provides broader and deeper logging of how users interact with the product.  All page views and all clicks on links, buttons, and inputs are now tracked throughout SCP.

Custom events beyond those in advanced search -- e.g. search within Study Overview, initializing visualizations like dot plots, igv.js, and Ideogram -- are now logged in Mixpanel as well.  This ports the handful of custom events logged to GA, and cleans up the logged objects for more consistency across the product.

Various JavaScript files now also conform to our new code style rules.

This satisfies the client-side code requirements of SCP-2326 and SCP-2382.

### CODE REVIEW

File: app/javascript/lib/scp-api-metrics.js [hunk 1]
```diff
@@ -84,7 +84,7 @@ export function logSearch(type, searchParams) {
     return
   }
 
-  const terms = searchParams.terms
+  const terms = searchParams.terms.split(' ')
```
Comment (bistline) [inline app/javascript/lib/scp-api-metrics.js:87]: Will this work if terms are passed URL encoded? (e.g. `%20` or `+`)?

File: app/javascript/packs/application.js [hunk 2]
```diff
@@ -30,38 +30,28 @@ import ClassicEditor from '@ckeditor/ckeditor5-build-classic'
...
-    // Logs only page views for faceted search UI
-    //
-    // If/when Mixpanel is extended beyond home page, remove study name from
-    // appPath in metrics-api.js at least for non-public studies to align with
-    // Terra on identifiable data that we want to omit from this logging.
-    logPageView()
+  // Logs only page views for faceted search UI
+  logPageView()
+
+  $(document).on('click', 'body', event => {
```
Comment (bistline) [inline app/javascript/packs/application.js:40]: I'm sure the answer is "yes", but this doesn't interfere with any other click handlers?  We have a lot...

Comment (eweitz) [reply app/javascript/packs/application.js:40]: Yup, I just re-verified several click handlers, including those added via `$(document).on('click'...`.

File: app/javascript/lib/scp-api-metrics.js (hunk 1 above)
Comment (eweitz) [reply app/javascript/lib/scp-api-metrics.js:87]: Yup!  I just verified that "foo+bar" and "baz%20moo" are both logged as two terms.

File: app/views/site/genome/_genome.html.erb [hunk 3]
```diff
@@ -13,10 +13,12 @@ See partials referenced below and scp-igv.js for larger bodies of code.
...
-      var myHeaders = new Headers({
-        'Authorization': 'Bearer ' + accessToken
-      });
-      arguments[1] = {headers: myHeaders};
+      if (arguments[0].includes('https://www.googleapis.com')) {
+        var myHeaders = new Headers({
+          'Authorization': 'Bearer ' + accessToken
+        });
+        arguments[1] = {headers: myHeaders};
+      }
```
Comment (devonbush) [inline app/views/site/genome/_genome.html.erb:21]: wow, I did not realize we were overriding window.fetch in SCP, although I think understand why.     Can you add a comment around this?

File: app/assets/javascripts/scp-dot-plot.js [hunk 4]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable no-undef */
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/assets/javascripts/scp-dot-plot.js]: Today I learned about `globals` in .eslintrc!  Done in 360b1c8.

File: app/assets/javascripts/scp-ideogram.js [hunk 5]
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable no-undef */
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/assets/javascripts/scp-ideogram.js]: Done in 360b1c8.

File: app/assets/javascripts/scp-igv.js [hunk 6]
```diff
@@ -1,3 +1,7 @@
+/* eslint-disable no-undef */
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/assets/javascripts/scp-igv.js]: Done in 360b1c8.

File: app/javascript/lib/metrics-api.js [hunk 7]
```diff
@@ -41,6 +41,9 @@ export function logPageView() {
 
 /** Log click on page.  Delegates to more element-specific loggers. */
 export function logClick(event) {
+  // Don't log programmatically-triggered events
+  if (typeof event.isTrigger !== 'undefined') return
+
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/javascript/lib/metrics-api.js:47]: Good question.  This does not affect such interactions -- I just verified that that `Enter` is still be logged (albeit as a click, like before).

This prevents logging events like those dispatched via jQuery's `trigger('click')`.  Morpheus does that upon initializing dot plots, for example.  I just added this clarification in the comment in 2414cba.

File: app/javascript/lib/metrics-api.js [hunk 8]
```diff
@@ -16,9 +16,9 @@ const defaultInit = {
 }
 
 const bardDomainsByEnv = {
-  'development': 'https://terra-bard-dev.appspot.com',
-  'staging': 'https://terra-bard-alpha.appspot.com',
-  'production': 'https://terra-bard-prod.appspot.com'
+  development: 'https://terra-bard-dev.appspot.com',
+  staging: 'https://terra-bard-alpha.appspot.com',
+  production: 'https://terra-bard-prod.appspot.com'
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/javascript/lib/metrics-api.js:21]: Worth doing, but I'm over the 15 minute clean-up limit.  I just put SCP-2237, the related ticket for this, into next sprint and assigned it to myself.  I'll ensure it gets done soon.

File: app/views/site/genome/_genome.html.erb (hunk 3 above)
Comment (eweitz) [reply app/views/site/genome/_genome.html.erb:24]: Done in e04553e.  I've also added a ticket for myself next sprint to remove this monkey patching of `fetch` -- SCP-2451.

File: app/javascript/lib/metrics-api.js [hunk 9]
```diff
@@ -137,6 +140,27 @@ export function logError(text) {
   log('error', props)
 }
 
+/**
+ * Removes study name from URL, as it might have identifying information.
+ * Terra UI omits workspace name in logs; this follows that precedent.
```
[the remark this answers is not in the document]
Comment (eweitz) [reply app/javascript/lib/metrics-api.js:146]: Worth noting!  Security mentioned this was prudent, but they didn't say it was required.

File: app/javascript/lib/metrics-api.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/lib/metrics-api.js:21]: awesome, thanks for sticking to the 15 min limit!

File: app/views/site/genome/_genome.html.erb [hunk 10]
```diff
@@ -11,6 +11,9 @@ See partials referenced below and scp-igv.js for larger bodies of code.
   <%= render partial: '/site/genome/ideogram' %>
   <script type="text/javascript" nonce="<%= content_security_policy_script_nonce %>">
     // Intercept requests to add bearer token, enabling direct load of files from GCS
+    // This was needed to load genome visualization data from GCS.
+    //
+    // TODO (SCP-2451): Remove monkey patching of fetch
```
Comment (devonbush) [inline app/views/site/genome/_genome.html.erb:16]: thanks for creating the issue

**Pull request merged**